src/nci_util.c

```diff
@@ -912,6 +1018,169 @@ nci_discovery_ntf_copy(
     }
 }
 
+NciModeParam*
+nci_mode_param_copy(
+    const NciModeParam* param,
+    NCI_MODE mode)
+{
+    if (G_LIKELY(param)) {
+        NciModeParam* copy = NULL;
+        gsize size = calculate_mode_param_fit_size(param, mode);
+        gsize copied = 0;
+
+        switch (mode) {
+        case NCI_MODE_ACTIVE_POLL_A:        /* fallthrough */
+        case NCI_MODE_PASSIVE_POLL_A:       /* fallthrough */
+        case NCI_MODE_ACTIVE_POLL_F:        /* fallthrough */
+        case NCI_MODE_PASSIVE_POLL_F:       /* fallthrough */
+        case NCI_MODE_PASSIVE_POLL_B:       /* fallthrough */
+        case NCI_MODE_ACTIVE_LISTEN_F:      /* fallthrough */
+        case NCI_MODE_PASSIVE_LISTEN_F:     /* fallthrough */
+            copy = (NciModeParam*) g_malloc0(size);
+            copied = nci_mode_param_inplace_copy(param, copy, mode);
```

Comment:
There's no need to allocate anything if we're not going to copy anything. If sizing and copying functions behave consistently, then the size could be checked for zero and then we know that the copying function will do the copying. Or perhaps the sizing function should always return at least `sizeof(NciModeParam)` and the copying function should always copy at least `sizeof(NciModeParam)` in which case there won't be any need to check anything. I like that even better.

Reply:
Or maybe not. Maybe NULL is a good representation of missing parameters. The the check is needed but only once.